Dalamud/Game/Addon/Lifecycle/AddonSetupHook.cs

```diff
@@ -0,0 +1,80 @@
+using System.Runtime.InteropServices;
+
+using Reloaded.Hooks.Definitions;
+
+namespace Dalamud.Game.Addon.Lifecycle;
+
+/// <summary>
+/// This class represents a callsite hook used to replace the address of the OnSetup function in r9.
+/// </summary>
+/// <typeparam name="T">Delegate signature for this hook.</typeparam>
+internal class AddonSetupHook<T> : IDisposable where T : Delegate
+{
+    private readonly Reloaded.Hooks.AsmHook asmHook;
+    
+    private T? detour;
+    private bool activated;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AddonSetupHook{T}"/> class.
+    /// </summary>
+    /// <param name="address">Address of the instruction to replace.</param>
+    /// <param name="detour">Delegate to invoke.</param>
+    internal AddonSetupHook(nint address, T detour)
+    {
+        this.detour = detour;
+
+        var detourPtr = Marshal.GetFunctionPointerForDelegate(this.detour);
+        var code = new[]
+        {
+            "use64",
+            $"mov r9, 0x{detourPtr:X8}",
+        };
+        
+        var opt = new AsmHookOptions
+        {
+            PreferRelativeJump = true,
+            Behaviour = Reloaded.Hooks.Definitions.Enums.AsmHookBehaviour.DoNotExecuteOriginal,
+            MaxOpcodeSize = 5,
+        };
+        
+        this.asmHook = new Reloaded.Hooks.AsmHook(code, (nuint)address, opt);
+    }
+
+    /// <summary>
+    /// Gets a value indicating whether or not the hook is enabled.
+    /// </summary>
+    public bool IsEnabled => this.asmHook.IsEnabled;
+    
+    /// <summary>
+    /// Starts intercepting a call to the function.
+    /// </summary>
+    public void Enable()
+    {
+        if (!this.activated)
+        {
+            this.activated = true;
+            this.asmHook.Activate();
+            return;
+        }
+        
+        this.asmHook.Enable();
+    }
+
+    /// <summary>
+    /// Stops intercepting a call to the function.
+    /// </summary>
+    public void Disable()
+    {
+        this.asmHook.Disable();
+    }
+
+    /// <summary>
+    /// Remove a hook from the current process.
+    /// </summary>
+    public void Dispose()
+    {
+        this.asmHook.Disable();
+        this.detour = null;
+    }
+}
```

Comment:
I was thinking of modifying CallHook and calling it something like CustomInstructionHook, where you pass in the instructions you want to write to the callsite via constructor arg or property setter, with a static field for what CallHook does currently being the default, this way we aren't duplicating a class in its entirety.

This is fine for now, as this is all internal I can clean it up later (after MSQ)

Reply:
Funny enough, I was thinking this myself. But then I was like: why not use Reloaded.Hooks.AsmHook directly instead of making a wrapper around it? But whatever.^^ 🤷‍♂️

Reply:
There will likely eventually be some kind of larger hook management system in Dalamud to potentially better help with cleanup, metrics, and other things. So for now I do plan to make a slightly friendlier wrapper.